### PR TITLE
Fix PaStA bugs and minor fixes

### PR DESCRIPTION
**Pull request by hechtlC**
### Prerequisites

- [X] I adhere to the coding conventions (described [here](https://github.com/se-passau/coronet/blob/master/CONTRIBUTING.md)) in my code.
- [X] I have updated the copyright headers of the files I have modified.
- [X] I have written appropriate commit messages, i.e., I have recorded the goal, the need, the needed changes, and the location of my code modifications for each commit. This includes also, e.g., referencing to relevant issues.
- [X] I have put signed-off tags in *all* commits.
- [X] I have updated the changelog file [NEWS.md](https://github.com/se-passau/coronet/blob/master/NEWS.md) appropriately.
- [X] The pull request is opened against the branch `dev`.

### Description

There are a couple of problems with the PaStA data in coronet which are fixed with this pull request. One problem is that the output of PaStA has a new, slightly different format. The file which contains the data is now called 'patch-groups' and at the end of the file there are new lines with all the commit hashes of a project that did not get matched in the following format: '=> hash'.

Another problem is that when merging the PaStA data to the mail and commit data, the revision set ids get duplicated in some cases. And there are cases where message ids or commit hashes that appear in the PaStA data do not appear in the mail and commit data. These problems are fixed with this pull request as well.

### Changelog

This is still missing and will be added both here and in the NEWS.md file once the pull request is reviewed.

#### Added
- Add function `cleanup.pasta.data` to remove wrong commit hashes and message ids from the PaStA data (1797e0324c39ad7b88dc22a14391340f4d26aea8, PR #189)

#### Fixed
- Fix reading of the PaStA data since the file format has changed (712bbafde3fb8f7b7c0fc847cb9c1838eb4cf86e, PR #189)
- Fix bug that duplicates revision set ids in the mail and commit data when merging the PaStA data and also copy-paste error when merging PaStA data to commit data (1797e0324c39ad7b88dc22a14391340f4d26aea8, PR #189)
- Fix bug that results in an error when there is a variable called 'c' in the R environment (de42eb24be131c261ccad7d807007f27d5559d68, PR #189)

### CODE REVIEW

File: util-data.R [hunk 1]
```diff
@@ -817,6 +829,32 @@ ProjectData = R6::R6Class("ProjectData",
...
+
+            ## remove message ids that don't appear in the mail data
+            if (!is.null(private$mails)) {
+                rev.id.contained = private$pasta[["revision.set.id"]] %in% private$mails[["revision.set.id"]]
+                private$pasta = private$pasta[rev.id.contained, ]
+            }
+
+            ## remove commit hashes that don't appear in the commit data
+            if (!is.null(private$commits)) {
+                pasta.commit.hashes = unlist(private$pasta[["commit.hash"]])
```
Comment (bockthom) [inline util-data.R:847]: Maybe you can do the `!is.na` check already for `pasta.commit.hashes`. Then we have fewer elements for doing the `%in%` checks and you can remove this NA check below for `commit.hashes.to.eliminate`.

File: util-data.R [hunk 2]
```diff
@@ -817,6 +829,32 @@ ProjectData = R6::R6Class("ProjectData",
...
+                commit.hashes.contained = unlist(private$pasta[["commit.hash"]]) %in% private$commits[["hash"]]
+                commit.hashes.to.eliminate = pasta.commit.hashes[!commit.hashes.contained]
+                commit.hashes.to.eliminate = commit.hashes.to.eliminate[!is.na(commit.hashes.to.eliminate)]
+                rows.to.remove = unlist(private$pasta["commit.hash"]) %in% commit.hashes.to.eliminate
```
Comment (bockthom) [inline util-data.R:851]: Please use the square bracket notation when accessing `pasta[["commit.hash"]]`.

File: util-data.R [hunk 3]
```diff
@@ -373,6 +383,8 @@ ProjectData = R6::R6Class("ProjectData",
                 private$update.pasta.commit.data()
             }
 
+            logging::logwarn("There might be PaStA data that does not appear in the mail or commit data.
+                              To clean this up you can call the function cleanup.pasta().")
```
Comment (bockthom) [inline util-data.R:388]: Please use single quotation marks around `'cleanup.pasta.data()'`. (See also comment below)

File: util-data.R [hunk 4]
```diff
@@ -817,6 +829,32 @@ ProjectData = R6::R6Class("ProjectData",
             }
         },
 
+        #' Remove lines in the PaStA data that contain
+        #' message ids or commit hashes that don't appear
+        #' in the commit or mail data.
+        cleanup.pasta = function() {
```
Comment (bockthom) [inline util-data.R:836]: To be consistent with the update function, we should call this function `cleanup.pasta.data`.

And is there a reason why you use three lines for the function documentation? I bet that two lines would be enough as you don't use the maximally allowed length of 120 characters per line :wink

**Review by bockthom — CHANGES_REQUESTED** [review]

**Comment by bockthom** [comment]
Somehow my overall review comment disappeared. I am sorry for that. Here is it again:

Thanks for this nice little changes @hechtlC. There are four minor issues which need to be fixed before we merge this PR.
In addition to that, please adjust the formatting of your commit messages as they violate our conventions of commit messages: There has to be a line break after 80 characters when writing detailed explanations in commit messages.
As you have to rebase your commits for that, you can also incorporate all the four other changes on the fly during rebasing :wink:

File: util-read.R [hunk 5]
```diff
@@ -430,7 +430,7 @@ PASTA.LIST.DATA.TYPES = c(
     "character", "character", "character"
 )
 
-#' Read and parse the PaStA data from the 'mbox-result' file.
+#' Read and parse the PaStA data from the 'patch-groups' file.
```
Comment (bockthom) [inline util-read.R:434]: Shouldn't this be also changed in the corresponding file name in our test suite?

File: util-data.R (hunk 1 above)
Comment (hechtlC) [reply util-data.R:847]: I have just tried this and for some reason this does not work as intended. So I will just keep it the way it is.

Comment (bockthom) [reply util-data.R:847]: Ok, then let's keep this as is (and save it for future improvements :wink:)

File: util-data.R (hunk 2 above)
Comment (hechtlC) [reply util-data.R]: Done. Thanks for catching that.

File: util-data.R (hunk 3 above)
Comment (hechtlC) [reply util-data.R]: Done.

File: util-data.R (hunk 4 above)
Comment (hechtlC) [reply util-data.R]: Done.

File: util-data.R [hunk 6]
```diff
@@ -817,6 +829,31 @@ ProjectData = R6::R6Class("ProjectData",
...
+            }
+
+            ## remove commit hashes that don't appear in the commit data
+            if (!is.null(private$commits)) {
+                pasta.commit.hashes = unlist(private$pasta[["commit.hash"]])
+                commit.hashes.contained = unlist(private$pasta[["commit.hash"]]) %in% private$commits[["hash"]]
+                commit.hashes.to.eliminate = pasta.commit.hashes[!commit.hashes.contained]
+                commit.hashes.to.eliminate = commit.hashes.to.eliminate[!is.na(commit.hashes.to.eliminate)]
+                rows.to.remove = unlist(private$pasta[["commit.hash"]]) %in% commit.hashes.to.eliminate
+                private$pasta = private$pasta[!rows.to.remove,]
```
Comment (bockthom) [inline util-data.R:857]: Here a space is missing after the comma. Sorry for not saying this earlier.

**Comment by bockthom** [comment]
Please don't forget to add the new bug fix regarding patch-stack filtering (d9fcd19) to the changelog file.

File: util-data.R (hunk 6 above)
Comment (hechtlC) [reply util-data.R]: Done.

**Review by bockthom — APPROVED** [review]
Thanks for all the PaStA-related fixes @hechtlC , looks good!
So I will merge right away.

**Pull request merged**